ddtrace/propagation/http.py

```diff
@@ -483,10 +492,214 @@ def _extract(headers):
+
+        # parse out values
+        if dd:
+            sampling_priority_ts = dd.get("s")
+            if sampling_priority_ts:
+                sampling_priority_ts = int(sampling_priority_ts)
+
+            origin = dd.get("o")
+            # need to convert from t. to _dd.p.
+            other_propagated_tags = {
+                re.sub("t.", "_dd.p.", k): v
+                for (k, v) in dd.items()
+                if (_TraceContext._is_valid_datadog_trace_tag_key(k))
+            }
+
+            return sampling_priority_ts, other_propagated_tags, origin
+        else:
+            log.debug("no dd list member in tracestate from incoming request: %r", ts)
+            return None
+
+    @staticmethod
+    def _get_sampling_priority(sampling_priority_tp, sampling_priority_ts):
+        # type: (int, int) -> int
+        """
+        1. If the traceparent sampled flag == 0, and the tracestate sampling priority is not present or greater than 0,
+        set the sampling priority field to 0.
+
+        2. If the traceparent sampled flag == 1, and the tracestate sampling priority is
+        not present or less than or equal to 0, update the sampling priority field to 1.
+
+        3. If the traceparent sampled flag == 1, and the tracestate sampling priority is greater than 0,
+        use the extracted sampling priority value.
+
+        4. If the traceparent sampled flag == 0, and the tracestate sampling priority is less than or equal to 0,
+        use the extracted sampling priority value.
+        """
+
+        if sampling_priority_tp == 0 and (not sampling_priority_ts or sampling_priority_ts >= 0):
+            sampling_priority = 0
+
+        elif sampling_priority_tp == 1 and (not sampling_priority_ts or sampling_priority_ts < 0):
+            sampling_priority = 1
+
+        elif sampling_priority_tp == 1 and sampling_priority_ts > 0:
+            sampling_priority = sampling_priority_ts
+
+        elif sampling_priority_tp == 0 and sampling_priority_ts < 0:
+            sampling_priority = sampling_priority_ts
+
+        return sampling_priority
+
+    @staticmethod
+    def _extract(headers):
+        # type: (Dict[str, str]) -> Optional[Context]
+
+        try:
+            tp = _extract_header_value(_POSSIBLE_HTTP_HEADER_TRACEPARENT, headers)
+            if tp is None:
+                return None
+            traceparent_values = _TraceContext._get_traceparent_values(tp)
+        except (ValueError, AssertionError):
+            log.exception("received invalid w3c traceparent: %s ", tp)
+            return None
+
+        trace_id, span_id, sampling_priority = traceparent_values
+        meta = {"traceparent": tp}
+
+        ts = _extract_header_value(_POSSIBLE_HTTP_HEADER_TRACESTATE, headers)
+        if ts:
+            # the value MUST contain only ASCII characters in the
+            # range of 0x20 to 0x7E except comma (,) and equal-sign (=)
+            if re.search(r"[^\x20-\x7E]+", ts):
```

Comment:
Question: Would we need a similar check for traceparent or is it guaranteed to be ascii? I guess we'll just raise an exception if we can't process the traceparent. Should we follow the same approach here? Attempt to parse the tracestate and then catch the exception if it fails?

Reply:
For traceparent I guess it could theoretically be anything, but yeah we'd just raise an exception for it because we're casting and converting the values. For tracestate I'm not sure it's as simple, since there can always be unexpected key and tag values, we won't get an error if for example someone were to add `t.€€€€€:not_ascii` in there.